### DISCUSSION TITLE
Code annotations with Prism highlighting

### DISCUSSION THREAD

**Discussion by ghost** [squidfunk/mkdocs-material#3122] [Ideas]
We are using Prism highlighting because we have been unable to persuade the Pygments team to accept a grammar for the [q programming language](https://code.kx.com/q/). Everything has worked smoothly, but if Pygments would support q, we would revert to minimise variation from the standard MM config. 

I find the new [code annotations feature](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#adding-annotations) does not work with Prism highlighting. Any prospect of that happening? (Or do you have any pull with the Pygments team?)

**Comment by facelessuser** [comment]
Even if Pygments doesn't support q, you could create your lexer and either pull request it, or just use it on the side. Pygments as a plugin system by which you can provide custom lexers.

**Reply by squidfunk** [reply]
As mentioned in the [documentation](https://squidfunk.github.io/mkdocs-material/reference/code-blocks/#fn:1), code annotations currently don't work with JavaScript-based syntax highlighters:

> Code annotations require syntax highlighting with Pygments – they're currently not compatible with JavaScript syntax highlighters. Support will be added at a later point, allowing to always place code annotations at the end of lines. ↩

It's theoretically possible to add support for this, but it's not something that's easy or fast to fix 😐 Maybe @facelessuser can elaborate on how to use a custom lexer with Pygments, which might be a better option.

**Reply by facelessuser** [reply]
I have a repo here that creates some very, very basic lexers. Feel free to use them to learn more. https://github.com/facelessuser/pymdown-lexers

**Reply by AjaniBilby** [reply]
Quick question: Why is the term "Javascript syntax highlighters" always being thrown around as if the only two big methods is Pygments or "Javascript based"?

Why not tmLanguage support? As that is the 'standard' used by IDEs, so that way language implementers get two birds with one stone?
I presume VSCode's implementation of tmLanguage support is a Javascript (typescript) based implementation, but TextMate and Sublime's appear to not be.

I just find it weird that there seems to be two competing standards, if you want syntax highlighting on a webpage you use Pygment, but if you want it in an IDE it's always tmLanguage...
One major advantage of tmLanguage over Pygments is that isn't not tied to a implementation language, such as Python, but then ofc you don't get the nice API bindings of a library like Pygments.

**Reply by squidfunk** [reply]
Interesting idea. You know how it is with standards. AFAIK Pygments is the only build-time solution supported by the Highlight extension (and its predecessor CodeHilite). Material for MkDocs uses them as upstream dependencies.

**Reply by AjaniBilby** [reply]
It's kind of ironic that you say that since I actually originally asked about this in mkdocs (https://github.com/mkdocs/mkdocs/issues/2828) - and they said it was a mkdocs-material issue...

**Reply by squidfunk** [reply]
Jup, code annotations are currently only offered by Material for MkDocs, but they rely on Highlight, which itself relies on Pygments to do the highlighting, thus upstream.

**Reply by AjaniBilby** [reply]
From that sentence alone I already see how much of a pain switching the syntax highlighter will likely be - even if there was a 1:1 drop-in replacement...

**Reply by AjaniBilby** [reply]
> Even if Pygments doesn't support q, you could create your lexer and either pull request it, or just use it on the side. Pygments as a plugin system by which you can provide custom lexers.
> \- @facelessuser 

You mentioned that Pygments allows easily implementation of custom lexers, which it does as seen [here ](https://pygments.org/docs/lexerdevelopment/#adding-and-testing-a-new-lexer) - however this doesn't mention how you'd then add it into the MkDocs workflow, and more importantly in a way that Github actions will also have access to.

It seems like we'd need to specify the custom lexer we'd want to import within Python directly, but as far as I know, there are no hooks available to run our own Python during the initialisation of MkDocs
```py
# For a lexer named CustomLexer
your_lexer = load_lexer_from_file(filename, **options)
```

**Reply by facelessuser** [reply]
You can implement some as I've done [here](https://github.com/facelessuser/pymdown-lexers). I then include them in my documentation requirements as seen [here](https://github.com/facelessuser/pymdown-extensions/blob/main/requirements/docs.txt#L2). They just get installed with my other requirements in GitHub [actions](https://github.com/facelessuser/pymdown-extensions/blob/main/.github/workflows/deploy.yml#L29) and are available when I build the docs.

**Reply by facelessuser** [reply]
I purposely do not make my custom extensions available on PyPI, but you could make your own custom ones available there if that is desirable.

**Reply by AjaniBilby** [reply]
I'm not very used to the Python eco-system, but from my understanding, you're basically NPM requiring your own package which is hosted in a GitHub instead of in the package manager. Which would then mean two things from my understanding:
  1) I would need to put my lexer(s) in a package, possibly their own repo
  2) As far as I understand this would only install the packages, it wouldn't then load them when MkDocs runs, or bind them into the Pygments eco system - so from what I can see it would have no affect on the MkDocs highlighting.

**Reply by facelessuser** [reply]
> you're basically NPM requiring your own package which is hosted in a GitHub instead of in the package manager

From a node perspective, this would kind of be the same, yes.

> I would need to put my lexer(s) in a package, possibly their own repo

Yes, that is at least what I've done.

> As far as I understand this would only install the packages, it wouldn't then load them when MkDocs runs, or bind them into the Pygments eco system.

Incorrect. If you do it properly, they are installed with entry points that Pygments recognizes and loads them automatically. https://github.com/facelessuser/pymdown-lexers/blob/master/setup.py#L6.

>  so from what I can see it would have no affect on the MkDocs highlighting

I'm pointing to an example that I literally use, so I can confirm it does in fact work. See here where I highlight CriticMarkup syntax: 
https://facelessuser.github.io/pymdown-extensions/extensions/critic/#usage (see the "Markdown" tab in the examples).
